### ISSUE TITLE
Calling initializer with Lambda info

### ISSUE THREAD

**Issue by jgriepentrog** [odedniv/safe-cfn-custom-resource#3]
Hello - I noticed that the initializer isn't called with the lambda info. You've done a lot of great work making this module, so I wanted to get your feedback before potentially implementing something in cfn-custom-resource and coming off as not respecting the work you've put in.

With cfn-custom-resource today, I commonly do a fair amount of initialization in my Lambda handler using the Lambda event (e.g. destructuring ResourceProperties, etc.) that's common to all operations and making certain variables available to them.

Since the initializer in safe-cfn-custom-resource isn't called with that, I'm either left with repeating similar logic in each operation's function or using a workaround function defined in the initializer and called from each operation's function. The 2nd option isn't too bad, but it would be nice to not have to do that.

I did manage to get this working in [my fork](https://github.com/zippadd/safe-cfn-custom-resource) if I move the initializer into the handler function and it passes the tests, but noticed it breaks the ability to get the resource/initError as you've intended for unit testing in option 2. I haven't found a solution to resolve this as the initializer was originally intended to be executed at initialization time, not execution time.

So, I see 3 options from here:
1. I could use some help finding a solution to somehow expose resource/initError and/or call the initializer with the lambda code in a different way to prevent a breaking change
2. Introduce a breaking change (I can submit a PR from my fork)
3. Decline entirely at which point I'll either workaround/continue as is or implement something in cfn-custom-resource (don't worry, you won't hurt my feelings with this option)

Thanks for reviewing and looking forward to your feedback!

**Comment by odedniv** [comment]
Nonsense John! If I didn't want public feedback on this library I would not have made it open source. The whole point of open source is to discuss and improve on projects!

I understand your point.

I want to emphasize that the point of the "initialization" function is to wrap your initialization in a try/catch so that if it fails (e.g one of your `require`s are missing) - CloudFormation won't hold the stack hostage for 1 hour. The initialization function is only called once during the loading of the Lambda container, and not during every invocation - this is by design.

The idea behind splitting the one handler into `create`/`update`/`delete` is to save you a switch/case and ensure that the right values are being sent to CloudFormation's API (with this library you should not call the API yourself). This is done by specifying a strict interface of return values as seen in the README.

Before making a change I want to offer 2 options that may suit you:

```javascript
// option 1:
module.exports.handler = require('safe-cfn-custom-resource')(() => {
  return {
    initialize(event, context) {
      // common initialization goes here
    },
    create(event, context) {
      this.initialize(event, context);
      // creation goes here
    },
    update(event, context) {
      this.initialize(event, context);
      // updating goes here
    },
    delete(event, context) {
      this.initialize(event, context);
      // deletion goes here
    },
  };
});

// option 2:
module.exports.handler = require('safe-cfn-custom-resource')(() => {
  function initialize(event, context) {
    // common initialization goes here
  }
  return {
    create(event, context) {
      initialize(event, context);
      // creation goes here
    },
    update(event, context) {
      initialize(event, context);
      // updating goes here
    },
    delete(event, context) {
      initialize(event, context);
      // deletion goes here
    },
  };
});
```

I often call `this.create(event, context)` from the `update` function in stateless custom resources (e.g ones that do not create anything physical, just add data or invoke something), maybe this fits your needs?

```javascript
module.exports.handler = require('safe-cfn-custom-resource')(() => {
  function initialize(event, context) {
    // common initialization goes here
  }
  return {
    create(event, context) {
      // main logic goes here
    },
    update(event, context) {
      this.create(event, context);
      // note that you may want to return something different than create does,
      // for example you may not want the id to change
      // (otherwise CFN will invoke the delete as well since it's resource replacement)
    },
    delete(event, context) {
      // deletion goes here
    },
  };
});
```

A final option that does require changing this package is having `safe-cfn-custom-resource` recognize that there's an `initialize` function (that is somewhat like a constructor of a class), and call it prior to calling `create`/`update`/`delete`, but I wonder if it's not too specific and wouldn't be clearer to just call the function like mentioned above.

What do you think?

**Comment by jgriepentrog** [comment]
Thanks Oded! 

That makes sense for the initializer function and why the split for the handler into the different functions. I do like it makes it clear that each case is covered along with saving a switch statement.

Option 1&2 are along the lines of what I was thinking when I mentioned the workaround function. It works as you've outlined, it just doesn't make any variables declared in it directly available in the create/update/delete functions due to scoping. Returning the variables in an object could be done to make them available through destructuring. It's an extra step is all.

Given the desire to keep the initializer as a one time load per container, it seems Option 1 is probably the best option.

I do like the thought of having the package call the initialize function if it exists and and passing its results on to the create/update/delete functions. It could be less clear in some ways, but I do like that it can make it more flexible and be clear about what create/update/delete require for input. Seems cleaner for unit testing.

Plus, you can opt out by not including an initialize function, in which case `event` and `context` are passed to the create/update/delete functions as before.

Here's what I'm thinking it would look like (assuming `safe-cfn-custom-resource` changes in place):
```javascript
module.exports.handler = require('safe-cfn-custom-resource')(() => {
  return {
    initialize(event, context) {
      /* Example initialize below */
      const initializeReturn = [];

      const {ResourceProperties, OldResourceProperties} = event;
      const {ValueToGetArn} = ResourceProperties;
      initializeReturn.push(/* Get arn using ValueToGetArn */)

      if (OldResourceProperties) {
        const {ValueToGetArn: OldValue} = OldResourceProperties;
        initializeReturn.push(/* Get arn using OldValue */)
      }

      return initializeReturn;
    },
    create(arn) {
      // creation goes here
    },
    update(newArn, oldArn) {
      // updating goes here
    },
    delete(arn) {
      // deletion goes here
    },
  };
});
```

Then for unit testing, 
```javascript
const { handler } = require('./index');
const event = { /* CloudFormation event */ };
const context = { /* Lambda context */ };
const arn = "/* a test arn */";
const oldArn = "/* a test old arn */";
handler.resource.initialize(event, context); /* <- makes sure this returns the correct info for the events */
handler.resource.create(arn); /* <- test directly with input, not worrying about the CF event */
handler.resource.update(arn, oldArn);
handler.resource.delete(arn);
handler(event, context); /* <- test everything together */
```

`safe-cfn-custom-resource` would use something like `Function.prototype.apply()` to execute the appropriate with the results of `initialize`.

My opinion though and I'm sure the mechanism could be improved/simplified. It did get a bit more complicated than I would like, especially for update/OldProperties. Would appreciate your thoughts. Thanks!

**Comment by jgriepentrog** [comment]
Good point on calling the create function from update by the way! I'll have to keep that in mind.

**Comment by odedniv** [comment]
Again 2 options :smile: 

**Option 1:** like option 1 above, but `initialize` can return stuff.
```javascript
module.exports.handler = require('safe-cfn-custom-resource')(() => {
  return {
    initialize(event, context) {
      // common initialization goes here
      return VALUE;
    },
    create(event, context) {
      let value = this.initialize(event, context);
      // creation goes here
    },
    update(event, context) {
      let value = this.initialize(event, context);
      // updating goes here
    },
    delete(event, context) {
      let value = this.initialize(event, context);
      // deletion goes here
    },
  };
});
```

**Option 2:** we can change the library to bind the methods so that this can be done (if it makes sense for your usecase)

```javascript
module.exports.handler = require('safe-cfn-custom-resource')(() => {
  return {
    initialize(event, context) {
      // common initialization goes here
      this.value = VALUE;
    },
    create(event, context) {
      this.initialize(event, context);
      // creation goes here, use this.value
    },
    update(event, context) {
      this.initialize(event, context);
      // updating goes here, use this.value
    },
    delete(event, context) {
      this.initialize(event, context);
      // deletion goes here, use this.value
    },
  };
});
```

What do you think of that?

**Comment by odedniv** [comment]
Another option that I think is better than the option 2 above is to allow the initializer to return a class (with the same function set on the `prototype` of it). When the library sees that the initializer returned an object with a `prototype` that contains the functions, it will construct it using the `new` keyword, supplying the event and context. This means instead of a custom `initialize` method you can actually have a JavaScript constructor and set stuff on `this`.

```javascript
module.exports.handler = require('safe-cfn-custom-resource')(() => {
  return class {
    constructor(event, context) {
      // common initialization goes here
      this.value = VALUE;
    }
    create(event, context) {
      // creation goes here, use this.value
    }
    update(event, context) {
      // updating goes here, use this.value
    }
    delete(event, context) {
      // deletion goes here, use this.value
    }
  };
});
```

Note that this way we wouldn't need commas between the functions :)
This also ensures separation between different requests (each one will have a different `this`).

**Comment by jgriepentrog** [comment]
Finally able to get back to this. Nice! I like the class option along with option 2. That closely mirrors what I'm doing with the common initialize method today by making those variables available in the create/update/delete functions.

I agree that class is a better solution. In addition to your points, seems much cleaner and less magical than option 2. Thanks! :)

**Comment by odedniv** [comment]
I'm glad you like it :smile: 

I may have some time to implement this over the weekend, if you need it sooner feel free to open a PR and I'll give my 2 cents.

The idea is to check if the `resource` object returned by the initializer contains create/update/delete, if it doesn't check if it has a `.prototype` that contains them. If it's the latter, use `new` passing the event/context and then on the returned object call create/update/delete.
